gtsfm/frontend/correspondence_generator/synthetic_correspondence_generator.py

```diff
@@ -0,0 +1,119 @@
+
+Authors: John Lambert
+"""
+import tempfile
+from typing import Dict, List, Tuple
+
+from dask.distributed import Client, Future
+import numpy as np
+import open3d
+
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.common.types import CAMERA_TYPE
+from gtsfm.frontend.correspondence_generator.correspondence_generator_base import CorrespondenceGeneratorBase
+from gtsfm.frontend.correspondence_generator.keypoint_aggregator.keypoint_aggregator_base import KeypointAggregatorBase
+from gtsfm.frontend.correspondence_generator.keypoint_aggregator.keypoint_aggregator_dedup import (
+    KeypointAggregatorDedup,
+)
+from gtsfm.frontend.correspondence_generator.keypoint_aggregator.keypoint_aggregator_unique import (
+    KeypointAggregatorUnique,
+)
+from gtsfm.loader.loader_base import LoaderBase
+from gtsfm.loader.tanks_and_temples_loader import TanksAndTemplesLoader
+
+
+class SyntheticCorrespondenceGenerator(CorrespondenceGeneratorBase):
+    """Pair-wise synthetic keypoint correspondence generator."""
+
+    def __init__(self, dataset_root: str, scene_name: str, deduplicate: bool = True) -> None:
+        """
+        Args:
+            dataset_root: Path to where Tanks & Temples dataset is stored.
+            scene_name: Name of scene from Tanks & Temples dataset.
+            deduplicate: Whether to de-duplicate with a single image the detections received from each image pair.
+        """
+        self._dataset_root = dataset_root
+        self._scene_name = scene_name
+        self._aggregator: KeypointAggregatorBase = (
+            KeypointAggregatorDedup() if deduplicate else KeypointAggregatorUnique()
+        )
+
+    def generate_correspondences(
+        self,
+        client: Client,
+        images: List[Future],
+        image_pairs: List[Tuple[int, int]],
+        num_sampled_3d_points: int = 500,
+    ) -> Tuple[List[Keypoints], Dict[Tuple[int, int], np.ndarray]]:
+        """Apply the correspondence generator to generate putative correspondences.
+
+        Args:
+            client: Dask client, used to execute the front-end as futures.
+            images: List of all images, as futures.
+            image_pairs: Indices of the pairs of images to estimate two-view pose and correspondences.
+
+        Returns:
+            List of keypoints, with one entry for each input image.
+            Putative correspondences as indices of keypoints (N,2), for pairs of images (i1,i2).
+        """
+        dataset_root = self._dataset_root
+        scene_name = self._scene_name
+
+        img_dir = f"{dataset_root}/{scene_name}"
+        poses_fpath = f"{dataset_root}/{scene_name}_COLMAP_SfM.log"
+        lidar_ply_fpath = f"{dataset_root}/{scene_name}.ply"
+        colmap_ply_fpath = f"{dataset_root}/{scene_name}_COLMAP.ply"
+        ply_alignment_fpath = f"{dataset_root}/{scene_name}_trans.txt"
+        bounding_polyhedron_json_fpath = f"{dataset_root}/{scene_name}.json"
+        loader = TanksAndTemplesLoader(
```

Comment:
this will have to be pinned to the machine with the input worker, right? Hence I was suggesting we create the correspondences on disk beforehand. But fine with it given the time constraint.